### PR TITLE
[#12] Provide possibility of implementing base, centralized configuration

### PR DESCRIPTION
**Pull request by wmaciejak**
**Solution for #12** 

Hi ya @jandudulski!
I finished my proposition of feature request related to inheriting the base configuration. It's really simple but I have a few doubts about the interface of `#configure` method. It's really annoying that there is repeating `configure` keyword but I wanted to stay consistent with dry-rb way. What do you think about it? 

```ruby
Gate::Rails.configure do
  configure do |config|
    config.messages_file = Rails.root.join("fixtures", "en.yml")

     def foo?(value)
      value == "FOO"
    end
  end
end
```

### CODE REVIEW

File: lib/gate/rails.rb [hunk 1]
```diff
@@ -22,7 +27,7 @@ def fetch_params_schema(schema_name)
       end
 
       def def_schema(&block)
-        @_schema = Dry::Validation.Params(&block)
+        @_schema = Dry::Validation.Params(BaseSchema, &block)
```
Comment (jandudulski) [inline lib/gate/rails.rb]: Whoa, didn't know that. Is it documented? :D

**Review by jandudulski — APPROVED** [review]
:heart:

**Comment by jandudulski** [comment]
Could you also please update readme?

File: lib/gate/rails.rb (hunk 1 above)
Comment (wmaciejak) [reply lib/gate/rails.rb]: Well, documentation don't know everything :D 
https://github.com/dry-rb/dry-validation/blob/master/lib/dry/validation.rb#L27

**Comment by wmaciejak** [comment]
@jandudulski done 👍

File: README.md [hunk 2]
```diff
@@ -89,6 +89,50 @@ class ExampleController < ActionController::Base
 end
 ```
 
+## Configuration
+
+Gate give possibility to create global, inherited configuration and separated setup in each one schema definition.
```
Comment (jandudulski) [inline README.md]: `s/give/gives`
`s/possibility/a possibility`

File: README.md [hunk 3]
```diff
@@ -89,6 +89,50 @@ class ExampleController < ActionController::Base
...
+
+### Inheriting base setup
+
+In order to configure the library according to whole application, create file under `config/initializers/gate.rb`. There is a possibility of creating base configuration, definining shared, custom predicates and a lot of more useful things. The full list of supported configuration options is consistent with available options in [dry-validation](https://dry-rb.org/gems/dry-validation/) gem.
```
Comment (jandudulski) [inline README.md]: `s/defininig/defining`
`s/file/a file`
`s/creating base/creating a base`
`s/lot of more/lot more`
`s/to whole/to the whole`

File: README.md [hunk 4]
```diff
@@ -89,6 +89,50 @@ class ExampleController < ActionController::Base
...
+    def foo?(value)
+      value == "FOO"
+    end
+  end
+end
+```
+
+### Separated setup
+
+Sometimes there is a requirement to have very specific configuration or behaviour in a one schema validation. The easies way is to doing it in special `configure` block.
```
Comment (jandudulski) [inline README.md]: `s/esies/easiest`
`s/very specific/a very specific`
`s/doing/do`
`s/special/a special`

File: lib/gate/rails.rb [hunk 5]
```diff
@@ -10,6 +10,11 @@ def self.included(base)
       base.send(:extend, ClassMethods)
     end
 
+    def self.configure(&block)
+      dynamic_class = "BaseSchema"
+      Object.const_set(dynamic_class, Class.new(Dry::Validation::Schema, &block))
```
Comment (jandudulski) [inline lib/gate/rails.rb]: I wonder what's the difference, and why not this way:

`BaseSchema = Class.new(Dry::Validation::Schema, &block)`

or even:

`BaseSchema = Dry::Validation::Schema.call(&block)`
or
`Object.const_set("BaseSchema", Dry::Validation::Schema.call(&block))`

?

Comment (wmaciejak) [reply lib/gate/rails.rb]: From your propositions only the last one is valid(and I will use it!). Why? We are defining constant in method body. In Ruby, there is no possibility to create constant if there is a possibility an executing this part of code more than once.

Comment (wmaciejak) [reply lib/gate/rails.rb]: But a few minutes ago I noticed that gate without base configuration doesn't work proper. I'm on it.

Comment (wmaciejak) [reply lib/gate/rails.rb]: Done @jandudulski

**Review by jandudulski — APPROVED** [review]

**Pull request merged**